File: src/com/aicp/extras/fragments/StatusBarClockSettings.java
/*
 * Copyright (C) 2018 Android Ice Cold Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


package com.aicp.extras.fragments;

import static android.view.DisplayCutout.BOUNDS_POSITION_LEFT;
import static android.view.DisplayCutout.BOUNDS_POSITION_RIGHT;

import android.app.AlertDialog;
import android.content.ContentResolver;
import android.content.DialogInterface;
import android.content.DialogInterface.OnCancelListener;
import android.os.Bundle;
import androidx.preference.Preference;
import androidx.preference.Preference.OnPreferenceChangeListener;
import android.provider.Settings;
import android.text.format.DateFormat;
import android.util.Log;
import android.view.Menu;
import android.view.View;
import android.view.Gravity;
import android.widget.EditText;
import android.widget.LinearLayout;

import com.aicp.extras.BaseSettingsFragment;
import com.aicp.extras.R;

import com.aicp.gear.preference.SystemSettingIntListPreference;
import com.aicp.gear.preference.SystemSettingListPreference;
import com.aicp.gear.util.DeviceUtils;
import com.android.internal.util.aicp.AicpUtils;

import java.util.Date;

public class StatusBarClockSettings extends BaseSettingsFragment implements OnPreferenceChangeListener {


    private static final String TAG = "StatusBarClockSettings";

    private static final String CLOCK_DATE_FORMAT = "status_bar_clock_date_format";
    private static final String CLOCK_POSITION = "status_bar_clock_position";

    public static final int CLOCK_DATE_STYLE_LOWERCASE = 1;
    public static final int CLOCK_DATE_STYLE_UPPERCASE = 2;
    private static final int CUSTOM_CLOCK_DATE_FORMAT_INDEX = 18;

    private SystemSettingIntListPreference mClockPosition;
    private SystemSettingListPreference mClockDateFormat;

    @Override
    protected int getPreferenceResource() {
        return R.xml.status_bar_clock;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ContentResolver resolver = getActivity().getContentResolver();

        mClockPosition = (SystemSettingIntListPreference) findPreference(CLOCK_POSITION);

        mClockDateFormat = (SystemSettingListPreference) findPreference(CLOCK_DATE_FORMAT);
        mClockDateFormat.setOnPreferenceChangeListener(this);
        if (mClockDateFormat.getValue() == null) {
            mClockDateFormat.setValue("EEE");
        }

        parseClockDateFormats();
    }

    @Override
    public void onResume() {
        super.onResume();

        final boolean hasNotch = AicpUtils.hasNotch(getActivity());
        final int notchType = DeviceUtils.getCutoutType(getActivity());
        Log.v(TAG,"notchType: " + notchType);

        // Adjust status bar preferences for RTL
        if (getResources().getConfiguration().getLayoutDirection() == View.LAYOUT_DIRECTION_RTL) {
            if (hasNotch && !(notchType == BOUNDS_POSITION_LEFT || notchType == BOUNDS_POSITION_RIGHT)) {
                mClockPosition.setEntries(R.array.clock_position_entries_notch_rtl);
                mClockPosition.setEntryValues(R.array.clock_position_values_notch_rtl);
            } else {
                mClockPosition.setEntries(R.array.clock_position_entries_rtl);
                mClockPosition.setEntryValues(R.array.clock_position_values_rtl);
            }
        } else if (hasNotch && !(notchType == BOUNDS_POSITION_LEFT || notchType == BOUNDS_POSITION_RIGHT)) {
            mClockPosition.setEntries(R.array.clock_position_entries_notch);
            mClockPosition.setEntryValues(R.array.clock_position_values_notch);
        }
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
        AlertDialog dialog;
        ContentResolver resolver = getActivity().getContentResolver();
        if (preference == mClockDateFormat) {
            int index = mClockDateFormat.findIndexOfValue((String) newValue);

            if (index == CUSTOM_CLOCK_DATE_FORMAT_INDEX) {
                AlertDialog.Builder alert = new AlertDialog.Builder(getActivity());
                alert.setTitle(R.string.clock_date_string_edittext_title);
                alert.setMessage(R.string.clock_date_string_edittext_summary);

                LinearLayout container = new LinearLayout(getActivity());
                container.setOrientation(LinearLayout.VERTICAL);
                LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,LinearLayout.LayoutParams.WRAP_CONTENT);
                lp.setMargins(55, 20, 55, 20);

                final EditText input = new EditText(getActivity());
                String oldText = Settings.System.getString(
                    resolver,
                    Settings.System.STATUS_BAR_CLOCK_DATE_FORMAT);
                if (oldText != null) {
                    input.setText(oldText);
                }
                input.setLayoutParams(lp);
                input.setGravity(android.view.Gravity.TOP| Gravity.START);
                container.addView(input);
                alert.setView(container);

                alert.setPositiveButton(R.string.menu_save, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialogInterface, int whichButton) {
                        String value = input.getText().toString();
                        if (value.equals("")) {
                            return;
                        }
                        Settings.System.putString(resolver,
                            Settings.System.STATUS_BAR_CLOCK_DATE_FORMAT, value);

                        return;
                    }
                });

                alert.setNegativeButton(R.string.menu_cancel,
                    new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialogInterface, int which) {
                        return;
                    }
                });
                dialog = alert.create();
                dialog.show();
            } else {
                if ((String) newValue != null) {
                    Settings.System.putString(resolver,
                        Settings.System.STATUS_BAR_CLOCK_DATE_FORMAT, (String) newValue);
                }
            }
            return true;
        }
        return false;
    }

    private void parseClockDateFormats() {
        String[] dateEntries = getResources().getStringArray(
                R.array.clock_date_format_entries_values);
        CharSequence parsedDateEntries[];
        parsedDateEntries = new String[dateEntries.length];
        Date now = new Date();

        int lastEntry = dateEntries.length - 1;
        int dateFormat = Settings.System.getInt(getActivity()
                .getContentResolver(), Settings.System.STATUS_BAR_CLOCK_DATE_STYLE, 0);
        for (int i = 0; i < dateEntries.length; i++) {
            if (i == lastEntry) {
                parsedDateEntries[i] = dateEntries[i];
            } else {
                String newDate;
                CharSequence dateString = DateFormat.format(dateEntries[i], now);
                if (dateFormat == CLOCK_DATE_STYLE_LOWERCASE) {
                    newDate = dateString.toString().toLowerCase();
                } else if (dateFormat == CLOCK_DATE_STYLE_UPPERCASE) {
                    newDate = dateString.toString().toUpperCase();
                } else {
                    newDate = dateString.toString();
                }

                parsedDateEntries[i] = newDate;
            }
        }
        mClockDateFormat.setEntries(parsedDateEntries);
    }
}

File: src/com/aicp/extras/fragments/StatusBarClockSettings.kt
/*
 * SPDX-FileCopyrightText: 2018-2026 Android Ice Cold Project
 * SPDX-License-Identifier: Apache-2.0
*/
package com.aicp.extras.fragments

import android.os.Bundle
import android.os.UserHandle
import android.provider.Settings
import android.text.format.DateFormat
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.preference.ListPreference
import androidx.preference.Preference
import com.aicp.extras.BaseSettingsFragment
import com.aicp.extras.R
import com.aicp.gear.preference.SystemSettingListPreference
import java.util.Date

class StatusBarClockSettings : BaseSettingsFragment(),
    Preference.OnPreferenceChangeListener {

    companion object {
        private const val TAG = "StatusBarClockSettings"

        private const val CLOCK_DATE_FORMAT = "status_bar_clock_date_format"
        private const val CLOCK_DATE_POSITION = "status_bar_clock_date_position"
        private const val CLOCK_DATE_DISPLAY = "status_bar_clock_date_display"
        private const val CLOCK_DATE_STYLE = "status_bar_clock_date_style"

        const val CLOCK_DATE_STYLE_LOWERCASE = 1
        const val CLOCK_DATE_STYLE_UPPERCASE = 2
        private const val CUSTOM_CLOCK_DATE_FORMAT_INDEX = 18
    }

    private lateinit var mClockDatePosition: SystemSettingListPreference
    private lateinit var mClockDateFormat: ListPreference
    private lateinit var mClockDateDisplay: SystemSettingListPreference
    private lateinit var mClockDateStyle: SystemSettingListPreference

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.status_bar_clock, rootKey)

        mClockDateDisplay = findPreference<SystemSettingListPreference>(CLOCK_DATE_DISPLAY)!!
        mClockDatePosition = findPreference<SystemSettingListPreference>(CLOCK_DATE_POSITION)!!
        mClockDateStyle = findPreference<SystemSettingListPreference>(CLOCK_DATE_STYLE)!!
        mClockDateFormat = findPreference<ListPreference>(CLOCK_DATE_FORMAT)!!

        mClockDateDisplay.onPreferenceChangeListener = this
        mClockDatePosition.onPreferenceChangeListener = this
        mClockDateStyle.onPreferenceChangeListener = this
        mClockDateFormat.onPreferenceChangeListener = this

        val dateDisplay = Settings.System.getIntForUser(
            requireContext().contentResolver,
            Settings.System.STATUS_BAR_CLOCK_DATE_DISPLAY,
            0,
            UserHandle.USER_CURRENT
        )

        mClockDatePosition.isEnabled = dateDisplay > 0
        mClockDateStyle.isEnabled = dateDisplay > 0
        mClockDateFormat.isEnabled = dateDisplay > 0

        if (mClockDateFormat.value == null) {
            mClockDateFormat.value = "EEE"
        }

        parseClockDateFormats()
    }

    override fun getPreferenceResource(): Int {
        return R.xml.status_bar_clock
    }

    override fun onPreferenceChange(preference: Preference, newValue: Any?): Boolean {
        when (preference) {
            mClockDateDisplay -> {
                val valInt = (newValue as String).toInt()
                val enabled = valInt != 0
                mClockDatePosition.isEnabled = enabled
                mClockDateStyle.isEnabled = enabled
                mClockDateFormat.isEnabled = enabled
                return true
            }
            mClockDatePosition, mClockDateStyle -> {
                parseClockDateFormats()
                return true
            }
            mClockDateFormat -> {
                val index = mClockDateFormat.findIndexOfValue(newValue as String)
                if (index == CUSTOM_CLOCK_DATE_FORMAT_INDEX) {
                    val alert = AlertDialog.Builder(requireActivity())
                        .setTitle(R.string.status_bar_date_string_edittext_title)
                        .setMessage(R.string.status_bar_date_string_edittext_summary)

                    val input = EditText(requireActivity())
                    val oldText = Settings.System.getString(
                        requireActivity().contentResolver,
                        Settings.System.STATUS_BAR_CLOCK_DATE_FORMAT
                    )
                    oldText?.let { input.setText(it) }
                    alert.setView(input)

                    alert.setPositiveButton(R.string.menu_save) { _, _ ->
                        val value = input.text.toString()
                        if (value.isNotEmpty()) {
                            Settings.System.putString(
                                requireActivity().contentResolver,
                                Settings.System.STATUS_BAR_CLOCK_DATE_FORMAT,
                                value
                            )
                        }
                    }

                    alert.setNegativeButton(R.string.cancel) { _, _ -> }
                    alert.create().show()
                } else {
                    Settings.System.putString(
                        requireActivity().contentResolver,
                        Settings.System.STATUS_BAR_CLOCK_DATE_FORMAT,
                        newValue as String
                    )
                }
                return true
            }
        }
        return false
    }

    override fun onResume() {
        super.onResume()
        val dateDisplay = Settings.System.getIntForUser(
            requireContext().contentResolver,
            Settings.System.STATUS_BAR_CLOCK_DATE_DISPLAY,
            0,
            UserHandle.USER_CURRENT
        )

        mClockDatePosition.isEnabled = dateDisplay > 0
        mClockDateStyle.isEnabled = dateDisplay > 0
        mClockDateFormat.isEnabled = dateDisplay > 0
        parseClockDateFormats()
    }

    private fun parseClockDateFormats() {
        val dateEntries = resources.getStringArray(R.array.status_bar_date_format_entries_values)
        val parsedDateEntries = Array<CharSequence>(dateEntries.size) { "" }
        val now = Date()
        val lastEntry = dateEntries.size - 1

        val dateFormat = Settings.System.getIntForUser(
            requireActivity().contentResolver,
            Settings.System.STATUS_BAR_CLOCK_DATE_STYLE,
            0,
            UserHandle.USER_CURRENT
        )

        for (i in dateEntries.indices) {
            parsedDateEntries[i] = if (i == lastEntry) {
                dateEntries[i]
            } else {
                val dateString = DateFormat.format(dateEntries[i], now).toString()
                when (dateFormat) {
                    CLOCK_DATE_STYLE_LOWERCASE -> dateString.lowercase()
                    CLOCK_DATE_STYLE_UPPERCASE -> dateString.uppercase()
                    else -> dateString
                }
            }
        }

        mClockDateFormat.entries = parsedDateEntries
    }
}
